Updates for better designer integration.

grokys opened the pull request:
## What does the pull request do?

I've been working on a new designer over at https://github.com/grokys/AvaloniaVS and while working on it, needed to make a few changes to Avalonia to make designing work better. I've included all the changes in this single PR; if you feel anything should be split out into a separate PR let me know.

The changes are:

### Strong name sign Avalonia.Remote.Protocol.

This may be a controversial change, but given that one of the main consumers of `Avalonia.Remote.Protocol` is VS and VS extensions need to be signed, it would make things a lot easier. This change should not be high impact because 1. `Avalonia.Remote.Protocol` has no dependencies and 2. strong name signing is ignored on .net core.

### Make `Styles` inherit from `AvaloniaObject`

Previously the `Design.PreviewWith` attached property had to be applied to the first `Style` within a `Styles` collection. By making the `Styles` collection an `AvaloniaObject` we can now add attached properties to it. This could also be useful elsewhere, but for the moment it just improves the UX of developing styles a little bit.

### Fixes some preview bitmap behavior

There were 2 problems with the bitmaps returned from the previewer process:

- No bitmap was returned if the control was empty. Instead return an empty bitmap.
- Bitmaps were sometimes returned for disposed `RemoteServerTopLevelImpl`s

### Ignore missing events in the designer

The designer would previously fail to display XAML which had event handlers assigned as the codebehind is not instantiated at design-time. 03e84dea7bf86d5133a9ed79891c24899e0d771a adds a way to override event lookup and uses that at design time to ignore event handlers.

### Update Portable.Xaml

Updates Portable.Xaml to include https://github.com/cwensley/Portable.Xaml/pull/140 which gives better line/column info for XAML errors.

### Allow null modifiers collection in input events

This one isn't really necessary, but it just makes the designer process not crash if one forgets to initialize `InputEventMessageBase.Messages`.

### Fix bad XAML in DesignWindowLoader message.

The XAML for the `Design.PreviewWith` message didn't have a `/` in the closing tag.

grokys committed:
Supply more detailed exception details.

So we can display better errors in a designer.

grokys committed:
Supply more detailed exception details.

So we can display better errors in a designer.

grokys committed:
Strong name sign Avalonia.Remote.Protocol.

It needs to be signed to be used in VS.

grokys committed:
Strong name sign Avalonia.Remote.Protocol.

It needs to be signed to be used in VS.

grokys committed:
Updated Portable.Xaml.

grokys committed:
Updated Portable.Xaml.

To get better line/column info for errors.

grokys committed:
Merge branch 'grokys/designer-updates' of https://github.com/AvaloniaUI/Avalonia into grokys/designer-updates

grokys committed:
Ignore missing events in the designer.

Do this by registering a special type converter for EventInfo, and add in the plumbing
for Portable.Xaml to use that type converter instead of its own (internal) `EventConverter`.

grokys committed:
Handle null modifiers collection.

grokys committed:
Return a frame even for an empty control.

If the user has a control with no content then we've still got a valid control
so return a frame with size 0.

grokys committed:
Don't invalidate a disposed TopLevelImpl.

grokys committed:
Fix bad XAML in DesignWindowLoader message.

grokys committed:
Make Styles inherit from AvaloniaObject.

This way we can add attached properties (such as `Design.PreviewWith`) to `Styles`.

grokys committed:
Make DesignWindowLoader expect PreviewWith on Styles.

grokys committed:
Merge branch 'master' into grokys/designer-updates

kekekeks commented on samples/ControlCatalog/SideBar.xaml:
```diff
@@ -1,6 +1,14 @@
 <Styles xmlns="https://github.com/avaloniaui"
         xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
         x:Class="ControlCatalog.SideBar">
+    <Design.PreviewWith>
```
I think it's not ignored in non-design mode, we might want to alter CompatibleXamlReader later or just wait for compiler to be ready.

grokys replied:
Yeah, I'd like to somehow make the XAML parser ignore `Design` attached properties. For the moment though, are you saying we should remove this?

kekekeks replied:
Since it's our control catalog, it shouldn't affect anything, so we can keep it.

kekekeks commented on src/Avalonia.Controls/Remote/Server/RemoteServerTopLevelImpl.cs:
```diff
@@ -301,8 +310,7 @@ protected void RenderIfNeeded()
                     return;
 
             }
-            if (ClientSize.Width < 1 || ClientSize.Height < 1)
```
What would happen if those are actually < 1? Or even < 0?

grokys replied:
If these are 0, then rendering is skipped over at line [276](https://github.com/AvaloniaUI/Avalonia/pull/2307/files/1f25265170c95ceb1f7ff2ba2b4ed35b565c84e9#diff-95da73f791f195c28fa00ea4a0696163R276).

Regarding <0 is that possible?

kekekeks commented on src/Avalonia.DesignerSupport/DesignerEventConverter.cs:
```diff
@@ -0,0 +1,87 @@
+                        }
+                    }
+
+                    // We want to ignore missing events in the designer, so if event handler
```
It's probably better to check for https://github.com/AvaloniaUI/Avalonia/blob/master/src/Markup/Avalonia.Markup.Xaml/PortableXaml/AvaloniaXamlSchemaContext.cs#L18 rather than rely on special converter being registered in DesignWindowLoader static constructor. Static constructor might be called in non-design mode and we'll end up ignoring missing event handlers.

grokys replied:
Good point about the static constructor. I'll move the registration out of that.

The reason I used a special converter however is that I didn't want to hard-code design-time behavior into `Avalonia.Markup.Xaml`. Don't you think this is a good idea? Would you prefer explicit design mode support to be added to `Avalonia.Markup.Xaml`?

kekekeks replied:
>Would you prefer explicit design mode support

It's already there - https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.DesignerSupport/DesignWindowLoader.cs#L21

grokys replied:
What I mean is: you say:

> rather than rely on special converter being registered in DesignWindowLoader static constructor.

Are you suggesting that instead of checking for a type converter in `AvaloniaMemberAttributeProvider`, that I should hard-code using the `DesignerEventConverter` in `Avalonia.Markup.Xaml` when `IsDesignMode == true`?

The reason I didn't do it like this is that we don't know how other designers will want to handle this case. By making it an extension point we're allowing the designer implementation to choose how to handle event handlers.

kekekeks replied:
I'm not sure what do you mean by "other designers". The upcoming compiler doesn't use converters for event subscriptions at all, it just emits `target.Event += context.RootObject.Method;`

grokys replied:
By "other designers" I just mean if someone comes along and has a great idea for a designer and wants to handle events differently. It's just feeling that we should expose extension points in something as general as `Avalonia.Markup.Xaml` rather than hard-coding designer behavior.

However I'm not tied to doing it like this and I'm unclear as to what you think I should do instead exactly. Please tell me how you think this should be done.

kekekeks replied:
I'd say, let's solve the problems when they arise. Portable.Xaml will be removed this year and compiler has completely different mechanism for event subscriptions.

grokys replied:
Ok, updated to do it like you suggested. I also added unit tests, which required using a non-shared `AvaloniaXamlSchemaContext` as otherwise the `IsDesignTime` setting interfered with other tests.

kekekeks replied:
>non-shared AvaloniaXamlSchemaContext

This could be a huge performance hit. Might be worth to keep two contexts (design and non-design one) for now.

grokys replied:
You're right of course. I saw that an instance of `IRuntimeTypeProvider` was being passed to `AvaloniaXamlSchemaContext` and assumed that that would do the caching of types/members etc. Looking more it seems not. Like you say keeping two contexts might be the best hack for now until the XAML compiler.

grokys replied:
Just pushed a fix for this.

kekekeks commented on src/Markup/Avalonia.Markup.Xaml/PortableXaml/AvaloniaMemberAttributeProvider.cs:
```diff
@@ -49,6 +49,18 @@ public object[] GetCustomAttributes(Type attributeType, bool inherit)
             }
+            else if (attributeType == typeof(TypeConverterAttribute) && _info is EventInfo)
+            {
+                // If a type converter for `EventInfo` is registered, then use that to convert
```
Again, better to check for DesignMode in AvaloniaXamlSchemaContext.

grokys replied:
See https://github.com/AvaloniaUI/Avalonia/pull/2307#discussion_r257517885

grokys committed:
Updated portable.xaml

grokys committed:
Fix failing test.

grokys committed:
Merge branch 'master' into grokys/designer-updates

grokys committed:
Move design-time handling of events...

...to `Avalonia.Markup.Xaml`. Also added tests: tests required a new instance of `AvaloniaXamlSchemaContext` for each `AvaloniaXamlLoader` as otherwise one test can affect other tests.

grokys committed:
Don't create a schema context each time.

The change in the previous commit to create a new `AvaloniaXamlSchemaContext` for each instance of `AvaloniaXamlLoader` had a negative performance effect. Instead, keep two instances of the schema context around: one for normal loading and one for design-time loading.

grokys committed:
Merge branch 'master' into grokys/designer-updates
